libs/langchain/langchain/callbacks/tracers/log_stream.py

```diff
@@ -0,0 +1,267 @@
+from __future__ import annotations
+
+import math
+import threading
+from typing import (
+    Any,
+    AsyncIterator,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    TypedDict,
+    Union,
+)
+from uuid import UUID
+
+import jsonpatch
+from anyio import create_memory_object_stream
+
+from langchain.callbacks.tracers.base import BaseTracer
+from langchain.callbacks.tracers.schemas import Run
+from langchain.schema.output import ChatGenerationChunk, GenerationChunk
+
+
+class LogEntry(TypedDict):
+    id: str
+    name: str
+    type: str
+    tags: List[str]
+    metadata: Dict[str, Any]
+    start_time: str
+
+    streamed_output_str: List[str]
+    final_output: Optional[Any]
+    end_time: Optional[str]
+
+
+class LogState(TypedDict):
+    id: str
+    entries: list[LogEntry]
+    streamed_output: List[Any]
+    final_output: Optional[Any]
```

Comment:
nit: would be useful to document the typedefs a bit more to explain intent / semantics

For example, not obvious why final output should be an optional (Suspected answer because of errors/cancellation)